hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromReferenceNode.java

```diff
@@ -137,4 +137,14 @@ protected boolean isFromElementUpdateOrDeleteRoot(FromElement element) {
 				|| getWalker().getStatementType() == HqlSqlTokenTypes.UPDATE;
 	}
 
+	/**
+	 * Returns a table names which are referenced by this node. If the tables
+	 * can not be determined it returns null.
```

Comment:
You say `If the tables can not be determined it returns null.` which IMHO would imply that for some reasons, we might not be able to determine the tables and in this case, we should disable the optimization. That's not what I saw in the rest of the code.

Is it just me misunderstanding or you had plans to take this into account?

If we don't have a special case for `null`, I think a collector pattern might be more efficient. It would avoid all these array initializations.

Reply:
It does not mean we are not able to deremine tables. Some nodes just does not refer any table. Tthen it returns null, It could also returns an empty array, there is no semantic difference between them.

Long answer:  for examle expression `tableAlias.prop` is translated to following AST:
```
DotNode
  +- IdentNode (tableAlias)
  \- IdentNode (prop)
```
The second `IdentNode `has no reference to persister and also no reference to fromElement (both are null). It is impossible to get a reference to table from that. But it does not mean I miss a reference I just get it from another node. For example here I get it from DotNode. So to sum it up I get information from a subset of nodes, not from all.